packages/aws-cdk/lib/api/accelerated-deployments.ts

```diff
@@ -0,0 +1,242 @@
+   * in case only one of them changed, like the key, and the Bucket stayed the same.
+   * However, that actually fails for old-style synthesis, which uses CFN Parameters!
+   * Because the names of the Parameters depend on the hash of the Asset,
+   * the Parameters used for the "old" values no longer exist in `assetParams` at this point,
+   * which means we don't have the correct values available to evaluate the CFN expression with.
+   * Fortunately, the diff will always include both the s3Bucket and s3Key parts of the Lambda's Code property,
+   * even if only one of them was actually changed,
+   * which means we don't need the "old" values at all, and we can safely initialize these with just `''`.
+   */
+  let s3Bucket = '', s3Key = '';
+  let foundCodeDifference = false;
+  // Make sure only the code in the Lambda function changed
+  const propertyUpdates = change.propertyUpdates;
+  for (const updatedPropName in propertyUpdates) {
+    const updatedProp = propertyUpdates[updatedPropName];
+    if (updatedProp.newValue === undefined) {
+      return false;
+    }
+    for (const newPropName in updatedProp.newValue) {
+      switch (newPropName) {
+        case 'S3Bucket':
+          foundCodeDifference = true;
+          s3Bucket = stringifyPotentialCfnExpression(updatedProp.newValue[newPropName], assetParamsWithEnv);
+          break;
+        case 'S3Key':
+          foundCodeDifference = true;
+          s3Key = stringifyPotentialCfnExpression(updatedProp.newValue[newPropName], assetParamsWithEnv);
+          break;
+        default:
+          return false;
+      }
+    }
+  }
+
+  return foundCodeDifference
+    ? {
+      s3Bucket,
+      s3Key,
+    }
+    : true;
+}
+
+/**
+ * For old-style synthesis which uses CFN Parameters,
+ * the Code properties can have the values of complex CFN expressions.
+ * For new-style synthesis of env-agnostic stacks,
+ * the Fn::Sub expression is used for the Asset bucket.
+ * Evaluate the CFN expressions to concrete string values which we need for the
+ * updateFunctionCode() service call.
+ */
+function stringifyPotentialCfnExpression(value: any, assetParamsWithEnv: { [key: string]: string }): string {
+  // if we already have a string, nothing to do
+  if (value == null || typeof value === 'string') {
+    return value;
+  }
+
+  // otherwise, we assume this is a CloudFormation expression that we need to evaluate
+  return evaluateCfn(value, assetParamsWithEnv);
+}
+
+function assetMetadataChanged(change: cfn_diff.ResourceDifference): boolean {
+  return !!change.newValue?.Metadata['aws:asset:path'];
+}
+
+export interface LambdaFunctionCode {
+  readonly s3Bucket: string;
+  readonly s3Key: string;
+}
+
+export interface LambdaFunctionResource {
+  readonly logicalId: string;
+  readonly physicalName?: string;
+  readonly code: LambdaFunctionCode;
+}
+
+async function applyAllAcceleratedChanges(
```

Comment:
Feels like there should be some abstraction here. i.e. the code that identifies accelerated changes should return an array of objects that encapsulate all the info for applying. So that this function just calls `await change.apply()` or something. Otherwise we split up the responsibility

Reply:
Introduced the abstraction, and implemented it with Lambda functions. Let me know if this is what you had in mind.